diff --git a/src/OpenSSL/SSL.py b/src/OpenSSL/SSL.py
index 12374b7..7f1b523 100644
--- a/src/OpenSSL/SSL.py
+++ b/src/OpenSSL/SSL.py
@@ -15,6 +15,8 @@ from OpenSSL._util import (
     path_bytes as _path_bytes,
     text_to_bytes_and_warn as _text_to_bytes_and_warn,
     no_zero_allocator as _no_zero_allocator,
+    wolfssl as _wolfssl,
+    wolfssl_debug as _wolfssl_debug
 )
 
 from OpenSSL.crypto import (
@@ -128,10 +130,12 @@ __all__ = [
 
 OPENSSL_VERSION_NUMBER = _lib.OPENSSL_VERSION_NUMBER
 SSLEAY_VERSION = _lib.SSLEAY_VERSION
-SSLEAY_CFLAGS = _lib.SSLEAY_CFLAGS
-SSLEAY_PLATFORM = _lib.SSLEAY_PLATFORM
-SSLEAY_DIR = _lib.SSLEAY_DIR
-SSLEAY_BUILT_ON = _lib.SSLEAY_BUILT_ON
+
+if not _wolfssl:
+    SSLEAY_CFLAGS = _lib.SSLEAY_CFLAGS
+    SSLEAY_PLATFORM = _lib.SSLEAY_PLATFORM
+    SSLEAY_DIR = _lib.SSLEAY_DIR
+    SSLEAY_BUILT_ON = _lib.SSLEAY_BUILT_ON
 
 SENT_SHUTDOWN = _lib.SSL_SENT_SHUTDOWN
 RECEIVED_SHUTDOWN = _lib.SSL_RECEIVED_SHUTDOWN
@@ -333,8 +337,22 @@ class _VerifyHelper(_CallbackExceptionHelper):
         @wraps(callback)
         def wrapper(ok, store_ctx):
             x509 = _lib.X509_STORE_CTX_get_current_cert(store_ctx)
-            _lib.X509_up_ref(x509)
-            cert = X509._from_raw_x509_ptr(x509)
+            if _wolfssl:
+                # _from_raw_x509_ptr has code to call X509_free on the X509
+                # object when it's time to do garbage collection. With wolfSSL,
+                # when that X509_free call happens, it's possible that the X509
+                # object has already been freed by FreeX509 (an internal
+                # function). That function doesn't care what the reference count
+                # is and does the free unconditionally. This causes problems
+                # when the garbage collector comes along and tries to free the
+                # already freed X509. The solution for the wolfSSL case is to
+                # duplicate the object rather than fiddling with the ref count.
+                x509_copy = _lib.X509_dup(x509)
+                cert = X509._from_raw_x509_ptr(x509_copy)
+            else:
+                _lib.X509_up_ref(x509)
+                cert = X509._from_raw_x509_ptr(x509)
+
             error_number = _lib.X509_STORE_CTX_get_error(store_ctx)
             error_depth = _lib.X509_STORE_CTX_get_error_depth(store_ctx)
 
@@ -647,8 +665,9 @@ def _make_requires(flag, error):
     return _requires_decorator
 
 
+alpn_criteria = True if _wolfssl else _lib.Cryptography_HAS_ALPN
 _requires_alpn = _make_requires(
-    _lib.Cryptography_HAS_ALPN, "ALPN not available"
+    alpn_criteria, "ALPN not available"
 )
 
 
@@ -1675,7 +1694,7 @@ class Connection(object):
             raise WantReadError()
         elif error == _lib.SSL_ERROR_WANT_WRITE:
             raise WantWriteError()
-        elif error == _lib.SSL_ERROR_ZERO_RETURN:
+        elif error == _lib.SSL_ERROR_ZERO_RETURN or (_wolfssl and error == _lib.SOCKET_PEER_CLOSED_E):
             raise ZeroReturnError()
         elif error == _lib.SSL_ERROR_WANT_X509_LOOKUP:
             # TODO: This is untested.
diff --git a/src/OpenSSL/_util.py b/src/OpenSSL/_util.py
index a3d5075..95f4e78 100644
--- a/src/OpenSSL/_util.py
+++ b/src/OpenSSL/_util.py
@@ -2,12 +2,19 @@ import os
 import sys
 import warnings
 
-from cryptography.hazmat.bindings.openssl.binding import Binding
-
-
-binding = Binding()
-ffi = binding.ffi
-lib = binding.lib
+wolfssl = True
+wolfssl_debug = False
+
+if wolfssl:
+    from wolfssl._ffi import ffi, lib
+    if wolfssl_debug:
+        lib.wolfSSL_Debugging_ON()
+else:
+    from cryptography.hazmat.bindings.openssl.binding import Binding
+
+    binding = Binding()
+    ffi = binding.ffi
+    lib = binding.lib
 
 
 # This is a special CFFI allocator that does not bother to zero its memory
diff --git a/src/OpenSSL/crypto.py b/src/OpenSSL/crypto.py
index a54fd54..6d00f5b 100644
--- a/src/OpenSSL/crypto.py
+++ b/src/OpenSSL/crypto.py
@@ -17,6 +17,7 @@ from OpenSSL._util import (
     UNSPECIFIED as _UNSPECIFIED,
     text_to_bytes_and_warn as _text_to_bytes_and_warn,
     make_assert as _make_assert,
+    wolfssl as _wolfssl
 )
 
 __all__ = [
@@ -177,9 +178,15 @@ def _get_asn1_time(timestamp):
     ):
         return _ffi.string(_lib.ASN1_STRING_data(string_timestamp))
     else:
-        generalized_timestamp = _ffi.new("ASN1_GENERALIZEDTIME**")
-        _lib.ASN1_TIME_to_generalizedtime(timestamp, generalized_timestamp)
-        if generalized_timestamp[0] == _ffi.NULL:
+        if not _wolfssl:
+            generalized_timestamp = _ffi.new("ASN1_GENERALIZEDTIME**")
+            _lib.ASN1_TIME_to_generalizedtime(timestamp, generalized_timestamp)
+            null_error = generalized_timestamp[0] == _ffi.NULL
+        else:
+            generalized_timestamp = _lib.ASN1_TIME_to_generalizedtime(timestamp, _ffi.NULL)
+            null_error = generalized_timestamp == _ffi.NULL
+
+        if null_error:
             # This may happen:
             #   - if timestamp was not an ASN1_TIME
             #   - if allocating memory for the ASN1_GENERALIZEDTIME failed
