--- httpd-2.4.46.orig/modules/ssl/ssl_engine_init.c	2021-11-18 12:40:42.903822554 -0500
+++ httpd-2.4.46.patched/modules/ssl/ssl_engine_init.c	2021-11-18 12:50:27.173308280 -0500
@@ -1390,6 +1390,9 @@
             }
         }
 
+/* We have not yet implemented SSL_CTX_check_private_key and
+ * SSL_CTX_get0_certificate() becuase FALCON is not supported by OpenSSL. */ 
+#if 0
         if (SSL_CTX_check_private_key(mctx->ssl_ctx) < 1) {
             ap_log_error(APLOG_MARK, APLOG_EMERG, 0, s, APLOGNO(02565)
                          "Certificate and private key %s from %s and %s "
@@ -1448,6 +1451,8 @@
         SSL_free(ssl);
 #endif
 
+#endif /* 0 */
+
         ap_log_error(APLOG_MARK, APLOG_INFO, 0, s, APLOGNO(02568)
                      "Certificate and private key %s configured from %s and %s",
                      key_id, certfile, keyfile);
@@ -1888,6 +1893,9 @@
     }
 #endif
 
+/* We have not yet implemented SSL_CTX_check_private_key() becuase FALCON is not
+ * supported by OpenSSL. */ 
+#if 0
     if (SSL_CTX_check_private_key(sc->server->ssl_ctx) != 1) {
         ap_log_error(APLOG_MARK, APLOG_EMERG, 0, s, APLOGNO(02572)
                      "Failed to configure at least one certificate and key "
@@ -1895,6 +1903,7 @@
         ssl_log_ssl_error(SSLLOG_MARK, APLOG_EMERG, s);
         return ssl_die(s);
     }
+#endif
 
 #if defined(HAVE_OCSP_STAPLING) && defined(SSL_CTRL_SET_CURRENT_CERT)
     /*
