Index: acinclude.m4
===================================================================
--- acinclude.m4	(revision 1883622)
+++ acinclude.m4	(working copy)
@@ -477,6 +477,100 @@
   fi
 ])
 
+dnl Check for wolfSSL
+ENABLED_WOLFSSL=no
+AC_MSG_NOTICE([----------------------------------------])
+AC_MSG_CHECKING([for wolfSSL])
+AC_ARG_WITH([wolfssl],
+  [AC_HELP_STRING([--with-wolfssl@<:@=DIR@:>@],
+    [Include wolfSSL support (default no, yes=/usr/local)]
+  )],
+  [
+    if test "x$withval" != "xno" ; then
+      dnl local variables
+      ap_wolfssl_mod_ldflags=""
+      ap_wolfssl_mod_cflags=""
+      ap_wolfssl_libs=""
+
+      dnl backup flags
+      saved_CPPFLAGS="$CPPFLAGS"
+      saved_LDFLAGS="$LDFLAGS"
+      saved_LIBS="$LIBS"
+
+      dnl determine wolfSSL location and flags
+      if test "x$withval" == "xyes" || test "x$withval" == "x" ; then
+          ap_wolfssl_mod_ldflags="-L/usr/local/lib"
+          ap_wolfssl_mod_cflags="-I/usr/local/include -I/usr/local/include/wolfssl"
+      else
+        if test -d "$withval/lib" ; then
+            ap_wolfssl_mod_ldflags="-L$withval/lib"
+        else
+            ap_wolfssl_mod_ldflags="-L$withval"
+        fi
+        if test -d "$withval/include"; then
+            ap_wolfssl_mod_cflags="-I$withval/include -I$withval/include/wolfssl"
+        else
+            ap_wolfssl_mod_cflags="-I$withval -I$withval/wolfssl"
+        fi
+      fi
+      ap_wolfssl_libs="-lwolfssl"
+      ap_wolfssl_mod_cflags="$ap_wolfssl_mod_cflags -DUSE_WOLFSSL"
+      
+      dnl test for wolfSSL
+      CPPFLAGS="$CPPFLAGS $ap_wolfssl_mod_cflags"
+      LDFLAGS="$LDFLAGS $ap_wolfssl_mod_ldflags"
+      LIBS="$LIBS $ap_wolfssl_libs"
+      AC_LINK_IFELSE([
+          AC_LANG_PROGRAM(
+          [[
+            #include <wolfssl/options.h>
+            #include <wolfssl/ssl.h>
+          ]], 
+          [[ 
+            wolfSSL_Init();
+          ]])
+        ],
+        [ wolfssl_linked=yes ],
+        [ wolfssl_linked=no ]
+      )
+
+      dnl restore flags
+      CPPFLAGS="$saved_CPPFLAGS"
+      LIBS="$saved_LIBS"
+      LDFLAGS="$saved_LDFLAGS"
+
+      if test "x$wolfssl_linked" == "xno" ; then
+        dnl handle library not found error
+        AC_MSG_ERROR([wolfSSL Library not found.
+            If it's already installed, specify its path using --with-wolfssl=/dir/])
+
+      else
+        ENABLED_WOLFSSL=yes
+
+        dnl make sure the SSL module is included
+        ac_cv_openssl=yes
+
+        AC_DEFINE(HAVE_WOLFSSL, 1, [Define if wolfSSL is available])
+
+        dnl add lib to module LDFLAGS and LIBS
+        MOD_LDFLAGS="$MOD_LDFLAGS $ap_wolfssl_libs"
+        LIBS="$LIBS $ap_wolfssl_libs"
+        APR_SETVAR(ab_LIBS, [$MOD_LDFLAGS])
+        APACHE_SUBST(ab_CFLAGS)
+        APACHE_SUBST(ab_LIBS)
+
+        dnl add CFLAGS and LDFLAGS to module and global
+        MOD_CFLAGS="$MOD_CFLAGS $ap_wolfssl_mod_cflags"
+        MOD_LDFLAGS="$MOD_LDFLAGS $ap_wolfssl_mod_ldflags"
+        CPPFLAGS="$CPPFLAGS $ap_wolfssl_mod_cflags"
+        LDFLAGS="$LDFLAGS $ap_wolfssl_mod_ldflags"
+      fi
+    fi
+  ]
+)
+AC_MSG_RESULT([$ENABLED_WOLFSSL])
+
+
 dnl
 dnl APACHE_CHECK_OPENSSL
 dnl
Index: modules/http2/h2_push.c
===================================================================
--- modules/http2/h2_push.c	(revision 1883622)
+++ modules/http2/h2_push.c	(working copy)
@@ -23,6 +23,7 @@
 #include <apr_time.h>
 
 #ifdef H2_OPENSSL
+#include <wolfssl/options.h>
 #include <openssl/sha.h>
 #endif
 
Index: modules/md/md_crypt.c
===================================================================
--- modules/md/md_crypt.c	(revision 1883622)
+++ modules/md/md_crypt.c	(working copy)
@@ -23,6 +23,7 @@
 #include <apr_file_io.h>
 #include <apr_strings.h>
 
+#include <wolfssl/options.h>
 #include <openssl/err.h>
 #include <openssl/evp.h>
 #include <openssl/pem.h>
@@ -64,7 +65,13 @@
 #endif
 #endif
 
+#if defined(USE_WOLFSSL)
 #ifndef OPENSSL_NO_CT
+#define OPENSSL_NO_CT
+#endif
+#endif
+
+#ifndef OPENSSL_NO_CT
 #include <openssl/ct.h>
 #endif
 
Index: modules/md/md_ocsp.c
===================================================================
--- modules/md/md_ocsp.c	(revision 1883622)
+++ modules/md/md_ocsp.c	(working copy)
@@ -26,6 +26,7 @@
 #include <apr_strings.h>
 #include <apr_thread_mutex.h>
 
+#include <wolfssl/options.h>
 #include <openssl/err.h>
 #include <openssl/evp.h>
 #include <openssl/ocsp.h>
Index: modules/md/mod_md.h
===================================================================
--- modules/md/mod_md.h	(revision 1883622)
+++ modules/md/mod_md.h	(working copy)
@@ -17,6 +17,7 @@
 #ifndef mod_md_mod_md_h
 #define mod_md_mod_md_h
 
+#include <wolfssl/options.h>
 #include <openssl/evp.h>
 #include <openssl/x509v3.h>
 
Index: modules/ssl/ssl_engine_init.c
===================================================================
--- modules/ssl/ssl_engine_init.c	(revision 1883622)
+++ modules/ssl/ssl_engine_init.c	(working copy)
@@ -93,6 +93,7 @@
 }
 #endif
 
+#ifndef HAVE_FIPS
 /*
  * Grab well-defined DH parameters from OpenSSL, see the BN_get_rfc*
  * functions in <openssl/bn.h> for all available primes.
@@ -133,16 +134,21 @@
     { BN_get_rfc2409_prime_1024, NULL, 0 }
 };
 
+#endif /* !HAVE_FIPS */
+
 static void init_dh_params(void)
 {
+#ifndef HAVE_FIPS
     unsigned n;
 
     for (n = 0; n < sizeof(dhparams)/sizeof(dhparams[0]); n++)
         dhparams[n].dh = make_dh_params(dhparams[n].prime);
+#endif
 }
 
 static void free_dh_params(void)
 {
+#ifndef HAVE_FIPS
     unsigned n;
 
     /* DH_free() is a noop for a NULL parameter, so these are harmless
@@ -152,6 +158,7 @@
         DH_free(dhparams[n].dh);
         dhparams[n].dh = NULL;
     }
+#endif
 }
 
 /* Hand out the same DH structure though once generated as we leak
@@ -163,6 +170,7 @@
  * to our copy. */
 DH *modssl_get_dh_params(unsigned keylen)
 {
+#ifndef HAVE_FIPS
     unsigned n;
 
     for (n = 0; n < sizeof(dhparams)/sizeof(dhparams[0]); n++)
@@ -169,6 +177,7 @@
         if (keylen >= dhparams[n].min)
             return dhparams[n].dh;
         
+#endif
     return NULL; /* impossible to reach. */
 }
 
@@ -216,6 +225,21 @@
 /*
  *  Per-module initialization
  */
+
+#ifdef HAVE_FIPS
+static void myFipsCb(int ok, int err, const char* hash)
+{
+    printf("in my Fips callback, ok = %d, err = %d\n", ok, err);
+    printf("message = %s\n", wc_GetErrorString(err));
+    printf("hash = %s\n", hash);
+
+    if (err == IN_CORE_FIPS_E) {
+        printf("In core integrity hash check failure, copy above hash\n");
+        printf("into verifyCore[] in fips_test.c and rebuild\n");
+    }
+}
+#endif /* HAVE_FIPS */
+
 apr_status_t ssl_init_Module(apr_pool_t *p, apr_pool_t *plog,
                              apr_pool_t *ptemp,
                              server_rec *base_server)
@@ -226,6 +250,11 @@
     apr_status_t rv;
     apr_array_header_t *pphrases;
 
+#ifdef HAVE_FIPS
+    wolfCrypt_SetCb_fips(myFipsCb);
+    //wolfSSL_Debugging_ON();
+#endif
+
     if (SSLeay() < MODSSL_LIBRARY_VERSION) {
         ap_log_error(APLOG_MARK, APLOG_WARNING, 0, base_server, APLOGNO(01882)
                      "Init: this version of mod_ssl was compiled against "
Index: modules/ssl/ssl_private.h
===================================================================
--- modules/ssl/ssl_private.h	(revision 1883622)
+++ modules/ssl/ssl_private.h	(working copy)
@@ -83,6 +83,7 @@
 #include "ap_expr.h"
 
 /* OpenSSL headers */
+#include <wolfssl/options.h>
 #include <openssl/opensslv.h>
 #if (OPENSSL_VERSION_NUMBER >= 0x10001000)
 /* must be defined before including ssl.h */
Index: modules/ssl/ssl_util_ssl.c
===================================================================
--- modules/ssl/ssl_util_ssl.c	(revision 1883622)
+++ modules/ssl/ssl_util_ssl.c	(working copy)
@@ -467,6 +467,7 @@
 DH *ssl_dh_GetParamFromFile(const char *file)
 {
     DH *dh = NULL;
+#ifndef HAVE_FIPS
     BIO *bio;
 
     if ((bio = BIO_new_file(file, "r")) == NULL)
@@ -473,6 +474,7 @@
         return NULL;
     dh = PEM_read_bio_DHparams(bio, NULL, NULL, NULL);
     BIO_free(bio);
+#endif
     return (dh);
 }
 
Index: support/ab.c
===================================================================
--- support/ab.c	(revision 1883622)
+++ support/ab.c	(working copy)
@@ -158,12 +158,13 @@
 
 #if defined(HAVE_OPENSSL)
 
+#include <wolfssl/options.h>
+#include <openssl/ssl.h>
 #include <openssl/rsa.h>
 #include <openssl/crypto.h>
 #include <openssl/x509.h>
 #include <openssl/pem.h>
 #include <openssl/err.h>
-#include <openssl/ssl.h>
 #include <openssl/rand.h>
 #define USE_SSL
 #define SK_NUM(x) sk_X509_num(x)
@@ -2554,7 +2555,9 @@
                     meth = TLSv1_2_client_method();
 #endif
                 } else if (strncasecmp(opt_arg, "TLS1", 4) == 0) {
+#if !defined(NO_OLD_TLS) && defined(WOLFSSL_ALLOW_TLSV10)
                     meth = TLSv1_client_method();
+#endif
                 }
 #else /* #if OPENSSL_VERSION_NUMBER < 0x10100000L */
                 meth = TLS_client_method();
