diff --git a/tests/wscript b/tests/wscript
index 685038a7a..7302a13dd 100644
--- a/tests/wscript
+++ b/tests/wscript
@@ -55,16 +55,6 @@ def build(ctx):
         "libntp/ymd2yd.c"
     ] + common_source
 
-    ctx.ntp_test(
-        features="c cprogram test",
-        target="test_libntp",
-        install_path=None,
-        defines=unity_config + ["TEST_LIBNTP=1"],
-        includes=[ctx.bldnode.parent.abspath(), "../include", "unity", "common"],
-        use="unity ntp parse M PTHREAD CRYPTO RT SOCKET NSL",
-        source=libntp_source,
-    )
-
     if ctx.env.REFCLOCK_GENERIC or ctx.env.REFCLOCK_TRIMBLE:
         # libparse available/required with generic and Trimble refclocks
 
@@ -103,17 +93,6 @@ def build(ctx):
         "ntpd/nts_extens.c",
     ]
 
-    ctx.ntp_test(
-        defines=unity_config + ["TEST_NTPD=1"],
-        features="c cprogram test",
-        includes=[ctx.bldnode.parent.abspath(), "../include", "unity", "../ntpd", "common", "../libaes_siv"],
-        install_path=None,
-        source=ntpd_source,
-        target="test_ntpd",
-        use="ntpd_lib libntpd_obj unity ntp aes_siv "
-            "M PTHREAD CRYPTO RT SOCKET NSL",
-    )
-
     testpylib.get_bld().mkdir()
 
     pypath = pylib.get_bld()
