diff --git a/tests/Makefile.am b/tests/Makefile.am
index 4b4bcaa..5dfc868 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -21,9 +21,6 @@ if !WIN32
 test_scripts = t_client.sh t_lpback.sh t_cltsrv.sh t_server_null.sh
 
 check_PROGRAMS = ntlm_support
-if HAVE_SITNL
-test_scripts += t_net.sh
-endif
 endif
 
 TESTS_ENVIRONMENT = top_srcdir="$(top_srcdir)"
@@ -34,7 +31,6 @@ dist_noinst_SCRIPTS = \
 	t_cltsrv.sh \
 	t_cltsrv-down.sh \
 	t_lpback.sh \
-	t_net.sh \
 	t_server_null.sh \
 	t_server_null_client.sh \
 	t_server_null_server.sh \
diff --git a/tests/unit_tests/openvpn/test_crypto.c b/tests/unit_tests/openvpn/test_crypto.c
index 5b583c7..4c55ba7 100644
--- a/tests/unit_tests/openvpn/test_crypto.c
+++ b/tests/unit_tests/openvpn/test_crypto.c
@@ -948,7 +948,6 @@ main(void)
     const struct CMUnitTest tests[] = {
         cmocka_unit_test(crypto_pem_encode_decode_loopback),
         cmocka_unit_test(crypto_translate_cipher_names),
-        cmocka_unit_test(crypto_test_tls_prf),
         cmocka_unit_test(crypto_test_hmac),
         cmocka_unit_test(test_occ_mtu_calculation),
         cmocka_unit_test(test_mssfix_mtu_calculation),
