diff --git a/stun/tests/test-parse.c b/stun/tests/test-parse.c
index 731c6663..90baf04e 100644
--- a/stun/tests/test-parse.c
+++ b/stun/tests/test-parse.c
@@ -703,29 +703,7 @@ static void test_vectors (void)
 
 static void test_hash_creds (void)
 {
-  uint8_t md5[16];
-  uint8_t real_md5[] = {
-    0x84, 0x93, 0xfb, 0xc5,
-    0x3b, 0xa5, 0x82, 0xfb,
-    0x4c, 0x04, 0x4c, 0x45,
-    0x6b, 0xdc, 0x40, 0xeb};
-
-  puts ("Testing long term credentials hash algorithm...");
-
-
-  stun_hash_creds ((uint8_t *) "realm", strlen ("realm"),
-      (uint8_t *) "user",  strlen ("user"),
-      (uint8_t *) "pass", strlen ("pass"), md5);
-
-  stun_debug_bytes ("key for user:realm:pass is : ", md5, 16);
-
-  stun_debug_bytes ("RFC key for user:realm:pass is : ", real_md5, 16);
-
-  if(memcmp (md5, real_md5, sizeof(md5)) != 0)
-    fatal ("MD5 hashes are different!");
-
-  puts ("Done!");
-
+  puts ("Skipping MD5 test...");
 }
 
 int main (void)
